File: backend/src/main/java/com/example/rentello/service/PricingService.java
package com.example.rentello.service;

import org.springframework.stereotype.Service;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.Month;
import java.util.Arrays;
import java.util.List;

@Service
public class PricingService {

    // Seasonal multipliers
    private static final BigDecimal SUMMER_MULTIPLIER = new BigDecimal("1.3"); // %30 increase
    private static final BigDecimal WINTER_MULTIPLIER = new BigDecimal("0.9"); // %10 decrease
    private static final BigDecimal SPRING_FALL_MULTIPLIER = new BigDecimal("1.1"); // %10 increase
    
    // Weekend multiplier
    private static final BigDecimal WEEKEND_MULTIPLIER = new BigDecimal("1.2"); // %20 increase
    
    // Holiday multiplier
    private static final BigDecimal HOLIDAY_MULTIPLIER = new BigDecimal("1.5"); // %50 increase
    
    // High demand period multiplier
    private static final BigDecimal HIGH_DEMAND_MULTIPLIER = new BigDecimal("1.4"); // %40 increase

    /**
     * Calculate dynamic price based on base rate and rental dates
     */
    public BigDecimal calculateDynamicPrice(BigDecimal baseRate, LocalDate startDate, LocalDate endDate) {
        if (baseRate == null || startDate == null || endDate == null) {
            return baseRate;
        }

        BigDecimal totalPrice = BigDecimal.ZERO;
        LocalDate currentDate = startDate;
        
        while (!currentDate.isAfter(endDate)) {
            BigDecimal dailyRate = calculateDailyRate(baseRate, currentDate);
            totalPrice = totalPrice.add(dailyRate);
            currentDate = currentDate.plusDays(1);
        }
        
        return totalPrice.setScale(2, RoundingMode.HALF_UP);
    }

    /**
     * Calculate daily rate with all multipliers applied
     */
    private BigDecimal calculateDailyRate(BigDecimal baseRate, LocalDate date) {
        BigDecimal multiplier = BigDecimal.ONE;
        
        // Apply seasonal multiplier
        multiplier = multiplier.multiply(getSeasonalMultiplier(date));
        
        // Apply weekend multiplier
        if (isWeekend(date)) {
            multiplier = multiplier.multiply(WEEKEND_MULTIPLIER);
        }
        
        // Apply holiday multiplier
        if (isHoliday(date)) {
            multiplier = multiplier.multiply(HOLIDAY_MULTIPLIER);
        }
        
        // Apply high demand period multiplier
        if (isHighDemandPeriod(date)) {
            multiplier = multiplier.multiply(HIGH_DEMAND_MULTIPLIER);
        }
        
        return baseRate.multiply(multiplier).setScale(2, RoundingMode.HALF_UP);
    }

    /**
     * Get seasonal multiplier based on date
     */
    private BigDecimal getSeasonalMultiplier(LocalDate date) {
        Month month = date.getMonth();
        
        // Summer (June, July, August)
        if (month == Month.JUNE || month == Month.JULY || month == Month.AUGUST) {
            return SUMMER_MULTIPLIER;
        }
        
        // Winter (December, January, February)
        if (month == Month.DECEMBER || month == Month.JANUARY || month == Month.FEBRUARY) {
            return WINTER_MULTIPLIER;
        }
        
        // Spring/Fall (March, April, May, September, October, November)
        return SPRING_FALL_MULTIPLIER;
    }

    /**
     * Check if date is weekend
     */
    private boolean isWeekend(LocalDate date) {
        DayOfWeek dayOfWeek = date.getDayOfWeek();
        return dayOfWeek == DayOfWeek.SATURDAY || dayOfWeek == DayOfWeek.SUNDAY;
    }

    /**
     * Check if date is a holiday
     */
    private boolean isHoliday(LocalDate date) {
        // Turkish holidays and international holidays
        List<LocalDate> holidays = getHolidays(date.getYear());
        return holidays.contains(date);
    }

    /**
     * Check if date is in high demand period
     */
    private boolean isHighDemandPeriod(LocalDate date) {
        Month month = date.getMonth();
        int day = date.getDayOfMonth();
        
        // Summer vacation period (July 1-31, August 1-15)
        if ((month == Month.JULY) || (month == Month.AUGUST && day <= 15)) {
            return true;
        }
        
        // New Year period (December 25-31, January 1-7)
        if ((month == Month.DECEMBER && day >= 25) || (month == Month.JANUARY && day <= 7)) {
            return true;
        }
        
        // Spring break period (April 15-30)
        if (month == Month.APRIL && day >= 15) {
            return true;
        }
        
        return false;
    }

    /**
     * Get list of holidays for a given year
     */
    private List<LocalDate> getHolidays(int year) {
        return Arrays.asList(
            // Fixed holidays
            LocalDate.of(year, 1, 1),   // New Year's Day
            LocalDate.of(year, 4, 23),  // National Sovereignty Day
            LocalDate.of(year, 5, 1),   // Labor Day
            LocalDate.of(year, 5, 19),  // Commemoration of Atatürk
            LocalDate.of(year, 7, 15),  // Democracy Day
            LocalDate.of(year, 8, 30),  // Victory Day
            LocalDate.of(year, 10, 29), // Republic Day
            LocalDate.of(year, 12, 25), // Christmas
            LocalDate.of(year, 12, 31)  // New Year's Eve
            // Note: Religious holidays (Eid) would need calculation based on lunar calendar
        );
    }

    /**
     * Get pricing breakdown for display
     */
    public PricingBreakdown getPricingBreakdown(BigDecimal baseRate, LocalDate startDate, LocalDate endDate) {
        PricingBreakdown breakdown = new PricingBreakdown();
        breakdown.setBaseRate(baseRate);
        breakdown.setStartDate(startDate);
        breakdown.setEndDate(endDate);
        
        BigDecimal totalPrice = BigDecimal.ZERO;
        int totalDays = 0;
        int weekendDays = 0;
        int holidayDays = 0;
        int highDemandDays = 0;
        
        LocalDate currentDate = startDate;
        while (!currentDate.isAfter(endDate)) {
            totalDays++;
            
            if (isWeekend(currentDate)) weekendDays++;
            if (isHoliday(currentDate)) holidayDays++;
            if (isHighDemandPeriod(currentDate)) highDemandDays++;
            
            BigDecimal dailyRate = calculateDailyRate(baseRate, currentDate);
            totalPrice = totalPrice.add(dailyRate);
            currentDate = currentDate.plusDays(1);
        }
        
        breakdown.setTotalDays(totalDays);
        breakdown.setWeekendDays(weekendDays);
        breakdown.setHolidayDays(holidayDays);
        breakdown.setHighDemandDays(highDemandDays);
        breakdown.setTotalPrice(totalPrice);
        breakdown.setAverageRate(totalPrice.divide(BigDecimal.valueOf(totalDays), 2, RoundingMode.HALF_UP));
        
        return breakdown;
    }

    /**
     * Pricing breakdown DTO
     */
    public static class PricingBreakdown {
        private BigDecimal baseRate;
        private LocalDate startDate;
        private LocalDate endDate;
        private int totalDays;
        private int weekendDays;
        private int holidayDays;
        private int highDemandDays;
        private BigDecimal totalPrice;
        private BigDecimal averageRate;

        // Getters and Setters
        public BigDecimal getBaseRate() { return baseRate; }
        public void setBaseRate(BigDecimal baseRate) { this.baseRate = baseRate; }
        
        public LocalDate getStartDate() { return startDate; }
        public void setStartDate(LocalDate startDate) { this.startDate = startDate; }
        
        public LocalDate getEndDate() { return endDate; }
        public void setEndDate(LocalDate endDate) { this.endDate = endDate; }
        
        public int getTotalDays() { return totalDays; }
        public void setTotalDays(int totalDays) { this.totalDays = totalDays; }
        
        public int getWeekendDays() { return weekendDays; }
        public void setWeekendDays(int weekendDays) { this.weekendDays = weekendDays; }
        
        public int getHolidayDays() { return holidayDays; }
        public void setHolidayDays(int holidayDays) { this.holidayDays = holidayDays; }
        
        public int getHighDemandDays() { return highDemandDays; }
        public void setHighDemandDays(int highDemandDays) { this.highDemandDays = highDemandDays; }
        
        public BigDecimal getTotalPrice() { return totalPrice; }
        public void setTotalPrice(BigDecimal totalPrice) { this.totalPrice = totalPrice; }
        
        public BigDecimal getAverageRate() { return averageRate; }
        public void setAverageRate(BigDecimal averageRate) { this.averageRate = averageRate; }
    }
} 

File: frontend/src/services/pricingService.ts
import api from './api';

export interface PriceCalculationRequest {
  vehicleId: number;
  startDate: string;
  endDate: string;
}

export interface PricingBreakdown {
  basePrice: number;
  weekendSurcharge: number;
  seasonalAdjustment: number;
  discountAmount: number;
  taxAmount: number;
  totalPrice: number;
  totalDays: number;
  averageRate: number;
  breakdown: any[];
}

export interface PriceCalculationResponse {
  vehicleId: number;
  baseRate: number;
  startDate: string;
  endDate: string;
  totalPrice: number;
  breakdown: PricingBreakdown;
}

export interface PricingMultipliers {
  seasonal: {
    summer: string;
    winter: string;
    springFall: string;
  };
  special: {
    weekend: string;
    holiday: string;
    highDemand: string;
  };
}

class PricingService {
  /**
   * Calculate dynamic price for a vehicle and date range
   */
  async calculatePrice(request: PriceCalculationRequest): Promise<PriceCalculationResponse> {
    const response = await api.post('/pricing/calculate', request);
    return response.data;
  }

  /**
   * Get pricing breakdown for display
   */
  async getPricingBreakdown(vehicleId: number, startDate: string, endDate: string): Promise<PricingBreakdown> {
    const response = await api.get('/pricing/breakdown', {
      params: { vehicleId, startDate, endDate }
    });
    return response.data;
  }

  /**
   * Get pricing multipliers information
   */
  async getPricingMultipliers(): Promise<PricingMultipliers> {
    const response = await api.get('/pricing/multipliers');
    return response.data;
  }

  /**
   * Format price for display
   */
  formatPrice(price: number): string {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 2
    }).format(price);
  }

  /**
   * Calculate days between two dates
   */
  calculateDays(startDate: string, endDate: string): number {
    const start = new Date(startDate);
    const end = new Date(endDate);
    const diffTime = Math.abs(end.getTime() - start.getTime());
    return Math.ceil(diffTime / (1000 * 60 * 60 * 24));
  }

  /**
   * Check if date is weekend
   */
  isWeekend(date: string): boolean {
    const d = new Date(date);
    const day = d.getDay();
    return day === 0 || day === 6; // Sunday = 0, Saturday = 6
  }

  /**
   * Check if date is holiday (basic implementation)
   */
  isHoliday(date: string): boolean {
    const d = new Date(date);
    const month = d.getMonth() + 1; // getMonth() returns 0-11
    const day = d.getDate();
    
    // Basic holiday check (can be expanded)
    const holidays = [
      { month: 1, day: 1 },   // New Year's Day
      { month: 4, day: 23 },  // National Sovereignty Day
      { month: 5, day: 1 },   // Labor Day
      { month: 5, day: 19 },  // Commemoration of Atatürk
      { month: 7, day: 15 },  // Democracy Day
      { month: 8, day: 30 },  // Victory Day
      { month: 10, day: 29 }, // Republic Day
      { month: 12, day: 25 }, // Christmas
      { month: 12, day: 31 }  // New Year's Eve
    ];
    
    return holidays.some(holiday => holiday.month === month && holiday.day === day);
  }

  /**
   * Get season for a date
   */
  getSeason(date: string): 'summer' | 'winter' | 'spring' | 'fall' {
    const d = new Date(date);
    const month = d.getMonth() + 1;
    
    if (month >= 6 && month <= 8) return 'summer';
    if (month >= 12 || month <= 2) return 'winter';
    if (month >= 3 && month <= 5) return 'spring';
    return 'fall';
  }

  /**
   * Get pricing explanation for a date range
   */
  getPricingExplanation(breakdown: PricingBreakdown): string[] {
    const explanations: string[] = [];
    
    explanations.push(`Base price: ${this.formatPrice(breakdown.basePrice)}`);
    explanations.push(`Total days: ${breakdown.totalDays}`);
    
    if (breakdown.weekendSurcharge > 0) {
      explanations.push(`Weekend surcharge: +${this.formatPrice(breakdown.weekendSurcharge)}`);
    }
    
    if (breakdown.seasonalAdjustment !== 0) {
      const sign = breakdown.seasonalAdjustment > 0 ? '+' : '';
      explanations.push(`Seasonal adjustment: ${sign}${this.formatPrice(breakdown.seasonalAdjustment)}`);
    }
    
    if (breakdown.discountAmount > 0) {
      explanations.push(`Discount: -${this.formatPrice(breakdown.discountAmount)}`);
    }
    
    explanations.push(`Tax (18%): ${this.formatPrice(breakdown.taxAmount)}`);
    explanations.push(`Average rate: ${this.formatPrice(breakdown.averageRate)} per day`);
    explanations.push(`Total price: ${this.formatPrice(breakdown.totalPrice)}`);
    
    return explanations;
  }
}

export default new PricingService(); 